🐛 BUG: Wrangler R2 GET requires local folder to exist

### Which Cloudflare product(s) does this pertain to?

R2

### What version of `Wrangler` are you using?

2.9.1

### What operating system are you using?

WSL

### Describe the Bug

If downloading a file from R2 that is stored in a prefix, that prefix needs to exist locally as a folder or the download fails.

```
$ wrangler r2 object get bucketname/foldername/filetodownload.zip
 ⛅️ wrangler 2.9.1
-------------------
Downloading "foldername/filetodownload.zip" from "bucketname".

✘ [ERROR] ENOENT: no such file or directory, open 'foldername/filetodownload.zip'
```
```
$ mkdir foldername
$ wrangler r2 object get bucketname/foldername/filetodownload.zip
 ⛅️ wrangler 2.9.1
-------------------
Downloading "foldername/filetodownload.zip" from "bucketname".
Download complete.
```

--- a/.changeset/perfect-students-pull.md
+++ b/.changeset/perfect-students-pull.md
@@ -1,0 +1,7 @@
+---
+"wrangler": patch
+---
+
+fix: automatically create required directories for `wrangler r2 object get`
+
+Previously, if you tried to use `wrangler r2 object get` with an object name containing a `/` or used the `--file` flag with a path containing a `/`, and the specified directory didn't exist, Wrangler would throw an `ENOENT` error. This change ensures Wrangler automatically creates required parent directories if they don't exist.

--- a/packages/wrangler/src/__tests__/r2.test.ts
+++ b/packages/wrangler/src/__tests__/r2.test.ts
@@ -295,6 +295,15 @@ describe("r2", () => {
 		`);
 		});
 
+		it("should download R2 object from bucket into directory", async () => {
+			await runWrangler(
+				`r2 object get bucketName-object-test/wormhole-img.png --file ./a/b/c/wormhole-img.png`
+			);
+			expect(fs.readFileSync("a/b/c/wormhole-img.png", "utf8")).toBe(
+				"wormhole-img.png"
+			);
+		});
+
 		it("should upload R2 object to bucket", async () => {
 			fs.writeFileSync("wormhole-img.png", "passageway");
 			await runWrangler(

--- a/packages/wrangler/src/r2/index.ts
+++ b/packages/wrangler/src/r2/index.ts
@@ -1,5 +1,6 @@
 import { Blob } from "node:buffer";
 import * as fs from "node:fs";
+import * as path from "node:path";
 import * as stream from "node:stream";
 import { ReadableStream } from "node:stream/web";
 import prettyBytes from "pretty-bytes";
@@ -109,7 +110,13 @@ export function r2(r2Yargs: CommonYargsArgv) {
 							logger.log(`Downloading "${key}" from "${fullBucketName}".`);
 						}
 
-						const output = file ? fs.createWriteStream(file) : process.stdout;
+						let output: stream.Writable;
+						if (file) {
+							fs.mkdirSync(path.dirname(file), { recursive: true });
+							output = fs.createWriteStream(file);
+						} else {
+							output = process.stdout;
+						}
 						if (objectGetYargs.local) {
 							await usingLocalBucket(
 								objectGetYargs.persistTo,

